vpr/src/analytical_place/full_legalizer.cpp

```diff
@@ -281,6 +291,710 @@ static LegalizationClusterId create_new_cluster(PackMoleculeId seed_molecule_id,
+            for (auto it = clusters.begin(); it != clusters.end() && !fit_in_a_neighbor; ) {
+                LegalizationClusterId cluster_id = *it;
+                if (!cluster_id.is_valid()) {
+                    ++it;
+                    continue;
+                }
+
+                // Get the cluster molecules and destroy the old cluster.
+                std::vector<PackMoleculeId> cluster_molecules = cluster_legalizer.get_cluster_molecules(cluster_id);
+                cluster_legalizer.destroy_cluster(cluster_id);
+
+                // Set the legalization strategy to speculative for fast try.
+                cluster_legalizer.set_legalization_strategy(ClusterLegalizationStrategy::SKIP_INTRA_LB_ROUTE);
+                
+                // Use the first molecule as seed to recreate the cluster.
+                PackMoleculeId seed_mol = cluster_molecules[0];
+                LegalizationClusterId new_cluster_id = create_new_cluster(seed_mol, prepacker_, cluster_legalizer, primitive_candidate_block_types); 
+
+                // Add remaining old molecules to the new cluster.
+                for (PackMoleculeId mol_id: cluster_molecules) {
+                    if (mol_id == seed_mol)
+                        continue;
+                    if (!cluster_legalizer.is_molecule_compatible(mol_id, new_cluster_id))
+                        continue;
+                    cluster_legalizer.add_mol_to_cluster(mol_id, new_cluster_id);
+                }
+
+                // Set the legalization strategy to full for adding new unclustered molecule.
+                // Also if recreated clusters if illegal, try to create with full strategy.
+                cluster_legalizer.set_legalization_strategy(ClusterLegalizationStrategy::FULL);
+
+                // If recreated cluster is illegal, try again with full strategy.
+                if (!cluster_legalizer.check_cluster_legality(new_cluster_id)) {
+                    cluster_legalizer.destroy_cluster(new_cluster_id);
+                    new_cluster_id = create_new_cluster(seed_mol, prepacker_, cluster_legalizer, primitive_candidate_block_types); 
+                    for (PackMoleculeId mol_id: cluster_molecules) {
+                        if (!cluster_legalizer.is_molecule_compatible(mol_id, new_cluster_id))
+                            continue;
+                        cluster_legalizer.add_mol_to_cluster(mol_id, new_cluster_id);
+                    }
+                }
+
+                // Lastly, try to add the new unclustered molecule to the recreated cluster.
+                if (cluster_legalizer.is_molecule_compatible(molecule_id, new_cluster_id)){
+                    e_block_pack_status pack_status = cluster_legalizer.add_mol_to_cluster(molecule_id, new_cluster_id);
+                    if (pack_status == e_block_pack_status::BLK_PASSED)
+                        fit_in_a_neighbor = true;
+                }
+
```

Comment:
Probably worth commenting that you open a cluster, try to add one thing to do it, then close it again.  This might cost CPU time if many molecules are packed in the same cluster in this pass, vs. just opening it once and adding them all.  Probably not worth updating though, as this only takes ~10% of flat recon time.